Add support for 'file' URI scheme to URL-based connectors

ptitzler opened the pull request:
Closes #2872
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our [contributor guidelines](https://github.com/elyra-ai/community/blob/main/contributing.md)
    1.1 Please follow the [7 rules for a great commit message](https://github.com/elyra-ai/community/blob/main/contributing.md#creating-a-pull-request)
  2. If the PR is unfinished, leave it as 'DRAFT', add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...' or use the 'status:Work in Progress' label.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If this PR involves UI changes, please provide a screen capture (before/after) for a faster review.
  6. Remember to add 'Fixes #ISSUE_NUMBER' (or any [valid keyword](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)) to the end of your message body to get the issue properly closed when the PR is merged.
  7. Set the proper milestone based on the target release for the issue.
  8. Consider whether any documentation need to be added or updated based on your changes.
-->

### What changes were proposed in this pull request?
The built-in URL catalog connector, Apache Airflow package connector, and the Apache Airflow provider package connectors now support input of a URL that uses the [`file` URI scheme](https://en.wikipedia.org/wiki/File_URI_scheme) in addition to the  already supported `HTTP` and `HTTPS` schemes. No changes to the UI or validation are required.

The behavior is as follows:
 - no changes if the user enters an HTTP-based URL
 - user enters a URL that identifies an existing local file name (e.g. `file://path/to/file`) -> HTTP status code 200
 - user enters a URL that identifies an existing local directory name (e.g. `file://i/am/a/dir`) -> HTTP status code 400 (invalid input)
 - user enters a URL that identifies a non-existing local file name (e.g. `file://no/such/file`) -> HTTP status code 404 (not found)
- URL connector: If a file-scheme URL is entered, the pipeline file is not portable because the complete URL is stored (`file://path/to/file`) just like it is done for the filesystem connector when bo base directory is specified.
- Airflow package connector: If a file-scheme URL is entered, the pipeline file is portable because the file path is not stored. (The stored metadata is identical to the metadata that is stored if an HTTP-based URL is entered. )
- Airflow provider package connector: If a file-scheme URL is entered, the pipeline file is portable because the file path is not stored. (The stored metadata is identical to the metadata that is stored if an HTTP-based URL is entered.)


<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor code that leads to class changes, showing the updated class hierarchy will help reviewers.
  2. If there is design documentation, please add the link.
-->

### How was this pull request tested?
- Reviewed the output of `make docs` (`user_guide/pipeline-components.html`)
- Added/updated server tests for new utility function/connectors
- Tested the following scenarios
   - URL connector
       - configure connector using valid `file://absolute/path/to/file.whl` URL
       - configure connector using invalid `file://absolute/path/to/no-such-file.whl` URL (specifying a file that does not exist)
       - configure connector using invalid `file://absolute/path/to/dir` URL (specifying an existing directory)
   - Apache Airflow package catalog connector
       - configure connector using valid `file://absolute/path/to/file.whl` URL
       - configure connector using invalid `file://absolute/path/to/no-such-file.whl` URL  (specifying a file that does not exist)
       - configure connector using invalid `file://absolute/path/to/dir` URL (specifying an existing directory)
   - Apache Airflow provider package catalog connector
       - configure connector using valid `file://absolute/path/to/file.whl` URL
       - configure connector using invalid `file://absolute/path/to/no-such-file.whl` URL  (specifying a file that does not exist)
       - configure connector using invalid `file://absolute/path/to/dir` URL (specifying an existing directory)
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly
including negative and positive cases if possible.

If it was tested in a way different from regular unit tests, please clarify how you tested step by step,
ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.

If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

kevin-bates reviewed (commented):
This looks great.  Is it possible to add a test or two for this?  Are there any doc adjustments necessary?

ptitzler commented:
> This looks great. Is it possible to add a test or two for this? Are there any doc adjustments necessary?

Done! I've added server tests for the connectors and the new utility library. I've also updated the documentation.

kiersten-stokes commented on docs/source/user_guide/pipeline-components.md:
````diff
@@ -395,36 +395,62 @@ Examples (CLI):
+   ```
+ - Local file URL
+   ```
+   ['file://local/path/to/component.yaml']
````
This is perhaps the biggest NIT of all time, but the connector will throw a 404 error if only 2 `/` are supplied

```suggestion
   ['file:///local/path/to/component.yaml']
```

kevin-bates replied:
> This is perhaps the biggest NIT of all time

Not at all, great catch!  The third slash is the start of the absolute path and I wonder if, for clarity, the example URI should instead be:
```
['file:///absolute/path/to/component.yaml']
```
to emphasize this is an absolute path to the "Local file".

ptitzler replied:
I changed all occurrences of `file:...` in the doc to `file:///absolute/path/to/...`

kiersten-stokes commented on docs/source/user_guide/pipeline-components.md:
````diff
@@ -395,36 +395,62 @@ Examples (CLI):
+   ```
+ - Local copy of a downloaded Apache Airflow package
+   ```
+   file://local/path/to/apache_airflow-1.10.15-py2.py3-none-any.whl
````
Same as above 

```suggestion
   file:///local/path/to/apache_airflow-1.10.15-py2.py3-none-any.whl
```

kiersten-stokes commented on docs/source/user_guide/pipeline-components.md:
````diff
@@ -395,36 +395,62 @@ Examples (CLI):
+
+ - Local copy of a downloaded provider package
+   ```
+   file://local/path/to/apache_airflow_providers_http-2.0.2-py3-none-any.whl
````
Last one 🙂 

```suggestion
   file:///local/path/to/apache_airflow_providers_http-2.0.2-py3-none-any.whl
```

kiersten-stokes reviewed (approved):
LGTM! Super valuable for the Airflow packages - I noticed a big difference in load time

kevin-bates reviewed (approved):
Aside from the minor doc updates/suggestions, this looks great.  Transport adapters are very cool!
